# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1711492645 -3600
#      Tue Aug 17 11:42:58 2021 +0200
# Node ID 1844e74a1c5039415255eb995d3da20b381667ea
# Parent  c93a98c49a211750aee26105c06449fc06f8ff09
Branding 8: Sign setup.exe and all .exe and .dll files.
* * *
Branding 8: Also sign uninstall\helper.exe
* * *
Branding 8: Sign setup.exe and all .exe and .dll files at the correct point in time. Exclude MS DLLs from signing.

diff --git a/browser/installer/windows/sign.py b/browser/installer/windows/sign.py
new file mode 100644
--- /dev/null
+++ b/browser/installer/windows/sign.py
@@ -0,0 +1,16 @@
+import subprocess, sys
+
+p = subprocess.Popen(["C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
+    "Get-ChildItem -Path ..\\..\\..\\dist\\bin -Filter *.exe -File | Where-Object { $_.Name -in @(\"betterbird.exe\", \"pingsender.exe\", \"plugin-container.exe\", \"rnp-cli.exe\", \"rnpkeys.exe\", \"WSEnable.exe\") } | ForEach-Object { Set-AuthenticodeSignature -FilePath ..\\..\\..\\dist\\bin\\$_ -Certificate (Get-ChildItem -Path Cert:\CurrentUser\My\ -CodeSigningCert) } "],
+    stdout=sys.stdout)
+p.communicate()
+
+p = subprocess.Popen(["C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
+    "Get-ChildItem -Path ..\\..\\..\\dist\\bin\\uninstall -Filter helper.exe -File | ForEach-Object { Set-AuthenticodeSignature -FilePath ..\\..\\..\\dist\\bin\\uninstall\\$_ -Certificate (Get-ChildItem -Path Cert:\CurrentUser\My\ -CodeSigningCert) } "],
+    stdout=sys.stdout)
+p.communicate()
+
+p = subprocess.Popen(["C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
+    "Get-ChildItem -Path ..\\..\\..\\dist\\bin -Filter *.dll -File | Where-Object { $_.Name -notlike \"vcruntime140*\" -and $_.Name -notlike \"msvcp140.*\" -and $_.Name -notlike \"api-ms-win*\" -and $_.Name -notlike \"d3dcompiler_47.*\" } | ForEach-Object { Set-AuthenticodeSignature -FilePath ..\\..\\..\\dist\\bin\\$_ -Certificate (Get-ChildItem -Path Cert:\CurrentUser\My\ -CodeSigningCert) } "],
+    stdout=sys.stdout)
+p.communicate()
diff --git a/browser/installer/windows/sign_installer.py b/browser/installer/windows/sign_installer.py
new file mode 100644
--- /dev/null
+++ b/browser/installer/windows/sign_installer.py
@@ -0,0 +1,6 @@
+import subprocess, sys
+
+p = subprocess.Popen(["C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
+    "Get-ChildItem -Path ..\\..\\..\\..\\dist\\install\\sea -Filter *installer.exe -File | ForEach-Object { Set-AuthenticodeSignature -FilePath ..\\..\\..\\..\\dist\\install\\sea\\$_ -Certificate (Get-ChildItem -Path Cert:\CurrentUser\My\ -CodeSigningCert) } "],
+    stdout=sys.stdout)
+p.communicate()
diff --git a/browser/installer/windows/sign_setup.py b/browser/installer/windows/sign_setup.py
new file mode 100644
--- /dev/null
+++ b/browser/installer/windows/sign_setup.py
@@ -0,0 +1,6 @@
+import subprocess, sys
+
+p = subprocess.Popen(["C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
+    "Set-AuthenticodeSignature -FilePath instgen\\setup.exe -Certificate (Get-ChildItem -Path Cert:\CurrentUser\My\ -CodeSigningCert)"],
+    stdout=sys.stdout)
+p.communicate()
diff --git a/toolkit/mozapps/installer/packager.mk b/toolkit/mozapps/installer/packager.mk
--- a/toolkit/mozapps/installer/packager.mk
+++ b/toolkit/mozapps/installer/packager.mk
@@ -131,8 +131,12 @@ make-package-internal: prepare-package m
 	$(call MAKE_PACKAGE,$(DIST))
 	echo $(PACKAGE) > $(ABS_DIST)/package_name.txt
 
 make-package: FORCE
+ifeq (WINNT,$(OS_ARCH))
+	@echo 'Signing *.exe and *.dll.'
+	$(PYTHON3) $(topsrcdir)/browser/installer/windows/sign.py
+endif
 	$(MAKE) make-package-internal
 ifeq (WINNT,$(OS_ARCH))
 ifeq ($(MOZ_PKG_FORMAT),ZIP)
 	$(MAKE) -C windows ZIP_IN='$(ABS_DIST)/$(PACKAGE)' installer
diff --git a/toolkit/mozapps/installer/windows/nsis/makensis.mk b/toolkit/mozapps/installer/windows/nsis/makensis.mk
--- a/toolkit/mozapps/installer/windows/nsis/makensis.mk
+++ b/toolkit/mozapps/installer/windows/nsis/makensis.mk
@@ -62,9 +62,11 @@ ifdef MOZ_STUB_INSTALLER
 endif
 
 ifdef ZIP_IN
 installer:: $(CONFIG_DIR)/setup.exe $(ZIP_IN)
-	@echo 'Packaging $(WIN32_INSTALLER_OUT).'
+	@echo 'Signing setup.exe.'
+	$(PYTHON3) $(topsrcdir)/browser/installer/windows/sign_setup.py
+	@echo 'Packaging $(ABS_DIST)/$(PKG_INST_PATH)$(PKG_INST_BASENAME).exe'
 	$(NSINSTALL) -D '$(ABS_DIST)/$(PKG_INST_PATH)'
 	$(PYTHON3) $(MOZILLA_DIR)/mach repackage installer \
 	  -o '$(ABS_DIST)/$(PKG_INST_PATH)$(PKG_INST_BASENAME).exe' \
 	  --package-name '$(MOZ_PKG_DIR)' \
@@ -72,8 +74,10 @@ installer:: $(CONFIG_DIR)/setup.exe $(ZI
 	  --tag $(topsrcdir)/$(MOZ_BUILD_APP)/installer/windows/app.tag \
 	  --setupexe $(CONFIG_DIR)/setup.exe \
 	  --sfx-stub $(SFX_MODULE) \
 	  $(USE_UPX)
+	@echo 'Signing installer.exe.'
+	$(PYTHON3) $(topsrcdir)/browser/installer/windows/sign_installer.py
 ifdef MOZ_STUB_INSTALLER
 	$(PYTHON3) $(MOZILLA_DIR)/mach repackage installer \
 	  -o '$(ABS_DIST)/$(PKG_INST_PATH)$(PKG_STUB_BASENAME).exe' \
 	  --tag $(topsrcdir)/browser/installer/windows/stub.tag \
